### ISSUE TITLE
Suggestion of `match_as_ref` does not compile

### ISSUE THREAD

**Issue by ordovicia** [rust-lang/rust-clippy#4437]
Version: `clippy 0.0.212 (e3cb40e 2019-06-25)`

```rust
#[derive(Debug)]
struct E {
    source: Option<ParseIntError>,
}

impl Error for E {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        // Clippy emits `match_as_ref` warning on this `match`
        // and suggests using `Option::as_ref()`.
        match self.source {
            Some(ref s) => Some(s),
            None => None,
        }
    //  ^ help: try this: `self.source.as_ref()`
    
    // ... but using `as_ref()` causes E0308: mismatched types.
        
    //  self.source.as_ref()
//      ^^^^^^^^^^^^^^^^^^^^ expected trait std::error::Error, found struct `std::num::ParseIntError`
// 
// = note: expected type `std::option::Option<&(dyn std::error::Error + 'static)>`
//            found type `std::option::Option<&std::num::ParseIntError>`
    }
}
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a78c4c13500d45711668f23c2d95e8ff)

**Comment by lzutao** [comment]
If you apply `as_ref()` suggestion and remove ref binding like this:
```diff
-            Some(ref s) => Some(s),
+            Some(s) => Some(s),
```
It will work.

**Comment by mati865** [comment]
Clippy suggests to remove whole `match` block which does not compile. Nonetheless seems like a Rust bug in accepting:
```rust
match self.source.as_ref() {
            Some(s) => Some(s),
            None => None,
        }
```
but rejecting `self.source.as_ref()` without match.

**Comment by lzutao** [comment]
Ah, I see it now.

**Comment by ordovicia** [comment]
Thanks for the comments, all.

I realized that this is a Rust's bug or at least weird behavior (rust-lang/rust#53716).
As a workaround, manual coersion `self.source.as_ref().map(|s| s as &dyn Error)` works well (it causes allow-by-default `trivial_casts` lint though).

Then it would be better to suggest using `map` like above when the returned type of `as_ref` is to be coerced.
